source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,297 @@
+
+{%fi /images/kubernetes-controllers-a-new-hope/kubernetes-controllers-a-new-hope-1.png, Plugin system, Plugin system %}
+
+This plugin system should have certain traits:
+
+* Plugins should be well isolated between them aka no plugin failure/bad behaviour should annoy other plugins
+* The host should be able to control and restrict what the plugin can do
+* Users should be able to develop plugins with different languages
+
+This plugin system may live in a _mega controller_, or it may live directly inside the _kube-controller-manager_, which is the process with all the controllers that manage the Kubernetes built-in building blocks:
+
+{%fi /images/kubernetes-controllers-a-new-hope/kubernetes-controllers-a-new-hope-2.png, kube-controller-manager including the plugin system, kube-controller-manager including the plugin system %}
+
+## Isolated, multi-language, lightweight plugin system? That sounds hard...
+
+The requirements of the plugin system are not trivial, but lucky for us WebAssembly comes in rescue!
+
+Wasm is an instruction set, targetable from every programming language, that runs in an isolated Virtual Machine. From [webassembly.org](https://webassembly.org/): _"WebAssembly (abbreviated Wasm) is a binary instruction format for a stack-based virtual machine. Wasm is designed as a portable compilation target for programming languages, enabling deployment on the web for client and server applications"_. That sounds like the perfect fit for our idea:
+
+* Because is isolated by default, we define piece by piece all the possible interactions between the plugins (controllers) and the host
+* Because it runs in a sandboxed virtual machine, it can easily isolate catastrophic failures and guarantee safe access to specific memory regions
+* Because it's a general purpose instruction set, more similar to an ISA of a modern CPU other than the JVM bytecode, most languages can target it (GC-enabled, dynamic/static typed, ...). 
+
+A list of supported languages is available [here](https://webassembly.org/getting-started/developers-guide/) and [here](https://stackoverflow.com/questions/43540878/what-languages-can-be-compiled-to-webassembly-wasm).
+
+Although WebAssembly sounds like a web, or better a browser technology, in the last years several people are experimenting this technology [outside the browser](https://webassembly.org/docs/use-cases/#outside-the-browser). Most notably, CloudFlare FaaS service called Workers [use Wasm](https://blog.cloudflare.com/introducing-wrangler-cli/) to run C/Rust code using Google's V8 isolations.
+
+With WebAssembly, our plugin system looks like:
+
+{%fi /images/kubernetes-controllers-a-new-hope/kubernetes-controllers-a-new-hope-3.png %}
+
+When we develop a controller, the compiled artifact is a binary Wasm module. A Wasm engine loads this module, eventually compiles it in AOT cases, and executes it.
```

Comment:
Describe AOT or leave it out if not necessary.

Reply:
i specified it a couple of lines above